subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -60,18 +61,20 @@
     private final FileHasher hasher;
     private final Interner<String> stringInterner;
     private final DefaultExcludes defaultExcludes;
+    private final DirectorySnapshotterStatistics.Collector collector;
 
-    public DirectorySnapshotter(FileHasher hasher, Interner<String> stringInterner, Collection<String> defaultExcludes) {
+    public DirectorySnapshotter(FileHasher hasher, Interner<String> stringInterner, Collection<String> defaultExcludes, DirectorySnapshotterStatistics.Collector collector) {
         this.hasher = hasher;
         this.stringInterner = stringInterner;
         this.defaultExcludes = new DefaultExcludes(defaultExcludes);
+        this.collector = collector;
     }
 
     public CompleteFileSystemLocationSnapshot snapshot(String absolutePath, @Nullable SnapshottingFilter.DirectoryWalkerPredicate predicate, final AtomicBoolean hasBeenFiltered) {
         try {
             Path rootPath = Paths.get(absolutePath);
             PathVisitor visitor = new PathVisitor(predicate, hasBeenFiltered, hasher, stringInterner, defaultExcludes);
-            Files.walkFileTree(rootPath, DONT_FOLLOW_SYMLINKS, Integer.MAX_VALUE, visitor);
+            Files.walkFileTree(rootPath, DONT_FOLLOW_SYMLINKS, Integer.MAX_VALUE, new StatisticsCollectingVisitor(collector, visitor));
```

Comment:
We call `walkFileTree` again when we follow symlinks. We would need to do the decoration there as well if we want to capture the directories we visit at the target of the symlink.

Reply:
Ah, and there's even a third place in `DefaultDirectoryWalker` that I missed. Thanks.

Reply:
I am not sure we should instrument `DefaultDirectoryWalker`, yet. It doesn't re-use any data from the VFS (at least not yet). And it will possibly make the numbers harder to understand. WDYT?

Reply:
Counting all `DirectorySnapshotter` visitations now, and none of `DefaultDirectoryWalker`. I'm in two minds about the latter, but don't feel strongly either way.

Reply:
Let's leave it as is and see if we need instrumentation of `DefaultDirectoryWalker` at some point.